core/src/main/java/overflowdb/Graph.java

```diff
@@ -0,0 +1,296 @@
+        Optional.of(new HeapUsageMonitor(config.getHeapPercentageThreshold(), referenceManager)) :
+        Optional.empty();
+  }
+
+  private void initElementCollections(OdbStorage storage) {
+    long start = System.currentTimeMillis();
+    final Set<Map.Entry<Long, byte[]>> serializedNodes = storage.allNodes();
+    logger.info("initializing " + serializedNodes.size() + " nodes from existing storage - this may take some time");
+    int importCount = 0;
+    long maxId = currentId.get();
+
+    final Iterator<Map.Entry<Long, byte[]>> serializedVertexIter = serializedNodes.iterator();
+    while (serializedVertexIter.hasNext()) {
+      final Map.Entry<Long, byte[]> entry = serializedVertexIter.next();
+      try {
+        final NodeRef nodeRef = storage.getNodeDeserializer().get().deserializeRef(entry.getValue());
+        nodes.add(nodeRef);
+        importCount++;
+        if (importCount % 131072 == 0) {
+          logger.debug("imported " + importCount + " elements - still running...");
+        }
+        if (nodeRef.id > maxId) maxId = nodeRef.id;
+      } catch (IOException e) {
+        throw new RuntimeException("error while initializing vertex from storage: id=" + entry.getKey(), e);
+      }
+    }
+
+    currentId.set(maxId + 1);
+    indexManager.initializeStoredIndices(storage);
+    long elapsedMillis = System.currentTimeMillis() - start;
+    logger.info("initialized " + this.toString() + " from existing storage in " + elapsedMillis + "ms");
+  }
+
+
+  ////////////// STRUCTURE API METHODS //////////////////
+
+  public Node addNode(final String label, final Object... keyValues) {
+    return addNode(currentId.incrementAndGet(), label, keyValues);
+  }
+
+  public Node addNode(final long id, final String label, final Object... keyValues) {
+    if (isClosed()) {
+      throw new IllegalStateException("cannot add more elements, graph is closed");
+    }
+    if (nodes.contains(id)) {
+      throw new IllegalArgumentException(String.format("Vertex with id already exists: %s", id));
+    }
+
+    currentId.set(Long.max(id, currentId.get()));
+    final NodeRef node = createNode(id, label, keyValues);
+    nodes.add(node);
+    return node;
+  }
+
+  private NodeRef createNode(final long idValue, final String label, final Object... keyValues) {
+    if (!nodeFactoryByLabel.containsKey(label)) {
+      throw new IllegalArgumentException("No NodeFactory for label=" + label + " available.");
+    }
+    final NodeFactory factory = nodeFactoryByLabel.get(label);
+    final NodeDb node = factory.createNode(this, idValue);
+    PropertyHelper.attachProperties(node, keyValues);
+    this.referenceManager.registerRef(node.ref);
+
+    return node.ref;
+  }
+
+  @Override
+  public String toString() {
+    return String.format("%s [%d nodes]", getClass().getSimpleName(), nodeCount());
+  }
+
+  /**
+   * if the config.graphLocation is set, data in the graph is persisted to that location.
+   */
+  @Override
+  public void close() {
```

Comment:
gosh, yes, and in java this will look super ugly because we need to have 5 nested try/finally blocks, don't we? it'll be something like below, just separated into smaller methods, or do you have a better idea?

```
  public void close() {
    this.closed = true;
    try {
      heapUsageMonitor.ifPresent(monitor -> monitor.close());
    } finally {
      try {
        if (config.getStorageLocation().isPresent()) {
          /* persist to disk */
          indexManager.storeIndexes(storage);
          referenceManager.clearAllReferences();
        }
      } finally {
        try {
          referenceManager.close();
        } finally {
          storage.close();
        }
      }
    }
  }
```

Reply:
actually I was only into having a single `try` with just `referenceManager` and `storage` being in `finally`. 
I assumed `indexManager` and `referenceManager` can only be problematic. Ideally we would have proper resource management but this is Java....

5 nested tries also works, although it is indeed ugly.